lib/npm-auditer.js

```diff
@@ -67,8 +67,11 @@ function reportAudit(npmAudit, config) {
   return npmAudit;
 }
 
-function runNpmAudit(callback) {
-  childProcess.exec('npm audit --json', (_error, stdout, stderr) => {
+function runNpmAudit(registry, callback) {
+  const command = registry
+    ? `npm audit --json --registry ${registry}`
```

Comment:
`exec` spawns a sub-shell, so I guess that you'll want to quote the registry. I'm not sure though, as that might also cause problems if the registry should happen to have quotes in it (technically it shouldn't - but the error message might not be great on improper values)? Maybe we should just use `spawn` instead.

Reply:
Can confirm that quoting the registry is OK when the values don't have quotes. The errors are quite good right now for NPM!

Here's an example of the message when providing a valid address with no registry (npm):

![image](https://user-images.githubusercontent.com/12538019/51810155-2ee7ba80-2274-11e9-8ea0-28a31e9002a5.png)

Here's another case with an error (npm):

![image](https://user-images.githubusercontent.com/12538019/51810232-9c93e680-2274-11e9-946e-c60be066bb4e.png)

And success using `child_process.exec` (npm):

![image](https://user-images.githubusercontent.com/12538019/51810546-88e97f80-2276-11e9-9d4b-a64764a9f660.png)

Potentially related note, it doesn't seem that Yarn considers the `--registry` flag appropriately.

`yarn audit --json` seems to accept the `--registry` flag:

![image](https://user-images.githubusercontent.com/12538019/51810304-07ddb880-2275-11e9-86ff-ad7c5dcac009.png)

However, providing a path with no registry it seems to use the default:

![image](https://user-images.githubusercontent.com/12538019/51810394-8b97a500-2275-11e9-83de-a3d8a7ea1925.png)

I don't have access to a non-NPM registry, so I can't properly test Yarn.

Considering that `npm audit`'s output is a single JSON object, there won't be any benefits streaming using `spawn` vs. buffering the output with `exec`. There may be performance gains using one or the other, not sure. One clear advantage of using `spawn` is to unify the internal API between Yarn and NPM by having them both use `spawn`. This will make #39 a lot easier.

Reply:
> Considering that `npm audit`'s output is a single JSON object, there won't be any benefits streaming using `spawn` vs. buffering the output with `exec`.

I was speaking more to making sure that the argument gets passed to the process correctly. For example, what happens if the registry contains `&`? That might only be a control character in the shell's interactive mode; I'm not sure. I think that we always want the registry argument to be passed along literally, which might be difficult to achieve if it's being processed by a shell first.

Reply:
Another note, not sure if/how to handle this, but what if someone's registry definition got compromised (perhaps env var injection), we might be executing arbitrary code.

If the path got compromised and changed to `https://registry.npmjs.org"; echo "hi" > "test.txt`, it would execute the `echo`. This definitely works for `exec` when introducing the extra literal quotes. Does `spawn` mitigate this, or can this be handled in some way (escape characters)? Is this even an issue?

Reply:
Okay, I am on board to migrating from `exec` to `spawn` for NPM (using the `cross-spawn` lib that we already use for Yarn). I think this should be done in a separate PR and then have this PR revised. I have had difficulties in the past with `spawn` for displaying inappropriate errors due to uncaught exceptions (or rejections) in a `Promise` (see #45). That issue should be addressed, along with the `spawn` migration, in the same PR. Some work can also be done for #39.